src/Service/Translation/Intl/IntlMessageVariationsProvider.php

```diff
@@ -0,0 +1,262 @@
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Originally forked from
+ * https://github.com/yiisoft/yii2/blob/2.0.15/framework/i18n/MessageFormatter.php
+ */
+
+namespace App\Service\Translation\Intl;
+
+use App\Constant\Common\Translation\Domain;
+use App\Exception\Translation\MessageFormatNotSupportedException;
+use App\Exception\Translation\MessagePatternIsInvalidException;
+use App\Model\Translation\Intl\IntlMessageVariation;
+use App\Model\Translation\Intl\IntlVariableMessage;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+use Symfony\Component\Translation\Translator;
+use Symfony\Contracts\Cache\ItemInterface;
+
+/**
+ * Class IntlMessageVariationsProvider
+ *
+ * It only supports the following message formats:
+ *  * plural formatting for english ('one' and 'other' selectors)
+ *  * select format
+ *  * simple parameters
+ *
+ * It does NOT support the ['apostrophe-friendly' syntax](https://php.net/MessageFormatter.formatMessage).
+ * Also messages that are working with the fallback implementation are not necessarily compatible with the
+ * PHP intl MessageFormatter so do not rely on the fallback if you are able to install intl extension somehow.
+ *
+ * @author Alexander Makarov <sam@rmcreative.ru>
+ * @author Carsten Brandt <mail@cebe.cc>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class IntlMessageVariationsProvider
+{
+    private Translator $translator;
+
+    private FilesystemAdapter $cache;
+
+    private IntlVariableMessageNormalizer $normalizer;
+
+    /**
+     * IntlMessageVariationsProvider constructor.
+     *
+     * @param Translator                    $translator
+     * @param FilesystemAdapter             $cache
+     * @param IntlVariableMessageNormalizer $normalizer
+     */
+    public function __construct(Translator $translator, FilesystemAdapter $cache, IntlVariableMessageNormalizer $normalizer)
+    {
+        $this->translator = $translator;
+        $this->cache = $cache;
+        $this->normalizer = $normalizer;
+    }
+
+    /**
+     * @param string      $id
+     * @param string      $domain
+     * @param string|null $locale
+     *
+     * @return IntlVariableMessage|null
+     *
+     * @throws MessagePatternIsInvalidException
+     * @throws \Psr\Cache\InvalidArgumentException
+     */
+    public function provide(string $id, string $domain = Domain::DEFAULT, string $locale = null): ?IntlVariableMessage
+    {
+        $catalogue = $this->translator->getCatalogue($locale);
+        if (!$catalogue->has($id, $domain)) {
+            return null;
+        }
+        $pattern = $catalogue->get($id, $domain);
+
+        return $this->cache->get(
+            sprintf('translation.intl_message_variations.%s', $id),
```

Comment:
Follow the class namespace `translation.intl.intl_message_variations_provider.%s`